File: beams/beams/custom_scripts/job_applicant/job_applicant.py
import frappe
from frappe import _
from frappe.model.document import Document
from frappe.utils import get_url, now_datetime
from frappe.utils import nowdate
from frappe.utils import get_url_to_form
from frappe.utils.password import encrypt
from frappe.model.naming import make_autoname
import os
from frappe.utils import today, getdate

@frappe.whitelist()
def autoname(doc, method):
	doc.name = make_autoname("JOB-APP-" + ".YYYY.-.####")

@frappe.whitelist()
def validate(doc, method):
	'''
		Method which trigger on validate of Job Applicant
	'''
	validate_job_applicant_details(doc)
	update_shortlisted_status_if_all_interviews_cleared(doc)

	if doc.status == 'Pending Document Upload' and doc.is_form_submitted:
		doc.status = 'Document Uploaded'

def validate_job_applicant_details(doc):
	'''
		Validate the Job Applicant against Job Opening requirements
	'''
	if doc.job_title:
		if frappe.db.exists('Job Opening', doc.job_title):
			job_opening_doc = frappe.get_doc('Job Opening', doc.job_title)

			# Fetching Minimum Experience
			expected_experience = float(job_opening_doc.min_experience) if job_opening_doc.min_experience else 0
			experience = float(doc.min_experience) if doc.min_experience else 0

def update_shortlisted_status_if_all_interviews_cleared(doc):
	'''
		Update Job Applicant status to 'Shortlisted from Interview'
		if all interview rounds are cleared and status is Interview Completed.
	'''
	if doc.status == 'Interview Completed':
		all_cleared = True
		for row in doc.applicant_interview_rounds:
			if row.interview_status != 'Cleared':
				all_cleared = False
				break

		if all_cleared:
			doc.status = 'Shortlisted from Interview'

@frappe.whitelist()
def get_existing_local_enquiry_report(doc_name):
	'''
		Create a Local Enquiry Report if it doesn't already exist
	'''
	ler_id = None
	# Check if a Local Enquiry Report already exists for the given Job Applicant
	if frappe.db.exists('Local Enquiry Report', { 'job_applicant':doc_name }):
		ler_id = frappe.db.get_value('Local Enquiry Report', { 'job_applicant':doc_name })
	return ler_id

@frappe.whitelist()
def create_and_return_report(job_applicant):
	'''
		Create a Local Enquiry Report, show an alert message, and return its name
	'''
	if frappe.db.exists('Job Applicant', job_applicant):
		ler_doc = frappe.new_doc('Local Enquiry Report')
		ler_doc.job_applicant = job_applicant
		ler_doc.save(ignore_permissions=True)
		frappe.msgprint('Local Enquiry Report Created: <a href="{0}">{1}</a>'.format(
			get_url_to_form(ler_doc.doctype, ler_doc.name),
			ler_doc.name
		),alert=True,indicator='green')
		return ler_doc.name
	return None

@frappe.whitelist()
def send_magic_link(applicant_id):
	'''
		Sends a unique magic link to the specified job applicant's email for document upload.
		Args:
			applicant_name (str): The name of the job applicant.
		Returns:
			None
	'''
	if frappe.db.exists('Job Applicant', applicant_id):
		email_id, applicant_name = frappe.db.get_value('Job Applicant', applicant_id, ['email_id', 'applicant_name'])
		link = generate_magic_link(applicant_id)
		if frappe.db.exists('Email Template', 'Job Applicant Follow Up'):
			template = frappe.get_doc('Email Template', 'Job Applicant Follow Up')
			subject = frappe.render_template(template.subject, {'applicant_name': applicant_name})
			response = frappe.render_template(template.response, {
				'applicant_name': applicant_name,
				'magic_link': link
			})
			frappe.sendmail(
				recipients=[email_id],
				subject=subject,
				message=response,
				delayed= False
			)
			frappe.msgprint(f'Magic link sent to {email_id}')
			frappe.db.set_value('Job Applicant', applicant_id, 'status', 'Pending Document Upload')
			return link
		else:
			frappe.msgprint('Email Template "Job Applicant Follow Up" does not exist.', alert=True)

@frappe.whitelist()
def generate_magic_link(applicant_id):
	'''
		Generates and returns a magic link URL for the specified job applicant
		Args:
			applicant_name (str): The name of the job applicant
		Returns:
			str: The generated magic link URL
	'''
	link = '{0}/job_application_upload/upload_doc?applicant_id='.format(get_url())
	if frappe.db.exists('Job Applicant', applicant_id):
		encrypted_link = encrypt(applicant_id)
		link += encrypted_link
	return link

@frappe.whitelist()
def set_interview_rounds(doc, method):
	'''
		Method to set Interview Rounds to Job Applicant from Job Opening
	'''
	if frappe.db.exists('Job Opening', doc.job_title):
		job_opening_doc = frappe.get_doc('Job Opening', doc.job_title)
		if job_opening_doc.interview_rounds:
			for round in job_opening_doc.interview_rounds:
				doc.append('applicant_interview_rounds', {
					'interview_round': round.interview_round
				})
			doc.save()

@frappe.whitelist()
def get_job_opening_location(job_opening):
	'''
		Method to get Location from Job Opening
	'''
	location = None
	if frappe.db.exists('Job Opening', job_opening):
		location = frappe.db.get_value('Job Opening', job_opening, 'preffered_location') or None
	return location


@frappe.whitelist()
def validate_unique_application(doc, method):
	'''
	Method to validate that an applicant can only apply for a job opening once using the same email ID.

	'''
	if doc.email_id and doc.job_title:
		existing_applicant = frappe.db.exists(
			"Job Applicant",
			{
				"email_id": doc.email_id,
				"job_title": doc.job_title,
			}
		)
		if existing_applicant and existing_applicant != doc.name:
			frappe.throw(
				_("The applicant with email ID {0} has already applied for the job opening {1}.")
				.format(doc.email_id, doc.job_title)
			)

def fetch_designation(doc, method):
	if doc.job_title:
		# Fetch the designation from the Job Opening
		designation = frappe.db.get_value('Job Opening', doc.job_title, 'designation')
		if designation:
			doc.designation = designation
		else:
			frappe.throw(f"Designation not found for the selected Job Opening: {doc.job_title}")

def fetch_department(doc, method):
	if doc.job_title:
		# Fetch the designation from the Job Opening
		department = frappe.db.get_value('Job Opening', doc.job_title, 'department')
		if department:
			doc.department = department
		else:
			frappe.throw(f"Department not found for the selected Job Opening: {doc.job_title}")

@frappe.whitelist()
def calculate_and_validate_age(doc, method=None):
	"""
		Calculate and validate age for a Job Applicant.
	"""

	if isinstance(doc, str):
		doc = frappe.parse_json(doc)

	date_of_birth = doc.get("date_of_birth")
	if not date_of_birth:
		return None

	dob = getdate(date_of_birth)
	today_date = getdate(today())

	if dob > today_date:
		frappe.throw(_("Date of Birth cannot be in the future."))

	age = today_date.year - dob.year - ((today_date.month, today_date.day) < (dob.month, dob.day))

	if age < 18:
		frappe.throw(_("Applicants must be at least 18 years old to apply."))
	doc.age = age
	return age

File: beams/beams/custom_scripts/job_applicant/job_applicant.js
frappe.ui.form.on('Job Applicant', {
	refresh: function (frm) {
		frm.remove_custom_button(__('Interview'), __('Create'));
		handle_custom_buttons(frm);
		frm.toggle_display('applicant_interview_rounds', !frm.is_new());
		const statuses = [
			'Document Uploaded', 'Open', 'Pending Document Upload', 'Shortlisted',
			'Local Enquiry Approved', 'Selected', 'Local Enquiry Started',
			'Local Enquiry Rejected', 'Local Enquiry Completed', 'Accepted', 'Training Completed', 'Compensation Proposal Accepted', 'Compensation Proposal Created'
		];

		if (statuses.includes(frm.doc.status)) {
			frm.remove_custom_button(__('Local Enquiry Report'), __('Create'));
		}

		if (frm.doc.status === 'Accepted' || frm.doc.status === 'Training Completed') {
			// Remove "View" button and its inner options
			frm.remove_custom_button('View');
			frm.page.remove_inner_button('Job Offer', 'View');

			// Remove "Send Magic Link" button
			frm.remove_custom_button('Send Magic Link');

			// Remove "Rejected" and "Hold" options under "Set Status" button
			frm.page.remove_inner_button('Rejected', 'Set Status');
			frm.page.remove_inner_button('Hold', 'Set Status');
			frm.remove_custom_button(__('Job Offer'), __('Create'));
		}

		const magic_link_statuses = [
			'Interview Completed', 'Local Enquiry Approved', 'Selected',
			'Compensation Proposal Created', 'Compensation Proposal Accepted',
			'Interview Scheduled', 'Interview Ongoing'
		];

		if (magic_link_statuses.includes(frm.doc.status)) {
			frm.remove_custom_button('Send Magic Link');
		}

		if (frm.doc.status === 'Compensation Proposal Created' || frm.doc.status === 'Compensation Proposal Accepted') {
			frm.page.remove_inner_button('Rejected', 'Set Status');
			frm.page.remove_inner_button('Hold', 'Set Status');
		}

		if (frm.doc.status === 'Interview Scheduled' || frm.doc.status === 'Interview Ongoing') {
			frm.page.remove_inner_button('Local Enquiry Report', 'Create');
		}
	},
	job_title: function(frm) {
		if (frm.doc.job_title) {
			frappe.db.get_value('Job Opening', frm.doc.job_title, 'department', function (r) {
				if (r && r.department) {
					frm.set_value('department', r.department);
				}
			});
		}
	},
	status: function (frm) {
		frm.trigger('refresh');
	},
	date_of_birth: function(frm) {
		if (frm.doc.date_of_birth) {
			frappe.call({
				method: "beams.beams.custom_scripts.job_applicant.job_applicant.calculate_and_validate_age",
				args: {
					doc: frm.doc
				},
				callback: function(r) {
					if (r.message) {
						frm.set_value("age", r.message);
					}
				}
			});
		}
	}
});

function handle_custom_buttons(frm) {
	if (!frm.is_new()) {
		if (frappe.user_roles.includes('HR Manager')) {
			// Handle "Appointment Letter" button for "Training Completed" status
			if (frm.doc.status === 'Training Completed') {
				frappe.db.get_value('Appointment Letter', { 'job_applicant': frm.doc.name }, 'name', function (result) {
					if (!result || !result.name) {
						frm.add_custom_button(__('Appointment Letter'), function () {
							frappe.new_doc('Appointment Letter', {
								job_applicant: frm.doc.name
							});
						}, __('Create'));
					}
				});
			}

			// Handle "Compensation Proposal" button for "Selected" status
			if (frm.doc.status === 'Selected') {
				frappe.db.get_value('Compensation Proposal', { 'job_applicant': frm.doc.name }, 'name', function (result) {
					if (!result || !result.name) {
						frm.add_custom_button(__('Compensation Proposal'), function () {
							frappe.new_doc('Compensation Proposal', {
								job_applicant: frm.doc.name
							});
						}, __('Create'));
					} else {
						frm.remove_custom_button(__('Compensation Proposal'), __('Create'));
					}
				});
			}

			// Show a "Local Enquiry Report (LER)" button in the Job Applicant form only when the status is "Shortlisted from Interview"
			if (frm.doc.status == 'Shortlisted from Interview') {
				frappe.call({
					method: 'beams.beams.custom_scripts.job_applicant.job_applicant.get_existing_local_enquiry_report',
					args: {
						doc_name: frm.doc.name
					},
					callback: function (response) {
						if (response.message) {
							frappe.db.get_doc('Local Enquiry Report', response.message).then(report => {
								frm.add_custom_button(__('Local Enquiry Report'), function () {
									frappe.set_route('Form', 'Local Enquiry Report', report.name);
								}, __('View'));
							});
						} else {
							frm.add_custom_button(__('Local Enquiry Report'), function () {
								frappe.call({
									method: 'beams.beams.custom_scripts.job_applicant.job_applicant.create_and_return_report',
									args: {
										job_applicant: frm.doc.name
									},
									callback: function (createResponse) {
										if (createResponse.message) {
											frm.add_custom_button(__('Local Enquiry Report'), function () {
												frappe.set_route('Form', 'Local Enquiry Report', createResponse.message);
											}, __('View'));
											frm.refresh();
										}
									}
								});
							}, __('Create'));
						}
					}
				});
			}

			if (frm.doc.status == 'Shortlisted' || frm.doc.status == 'Pending Document Upload') {
				frm.add_custom_button(__('Send Magic Link'), function () {
					frappe.confirm('Are you sure you want to send the magic link to the candidate?', function () {
						frappe.call({
							method: 'beams.beams.custom_scripts.job_applicant.job_applicant.send_magic_link',
							args: {
								applicant_id: frm.doc.name
							},
							callback: function (r) {
								if (r.message) {
									navigator.clipboard.writeText(r.message).then(function () {
										frappe.show_alert(__('Magic Link copied to clipboard!'));
									}).catch(function (err) {
										frappe.show_alert(__('Failed to copy Magic Link to clipboard'));
									});
								}
							}
						});
					});
				});
			}

			if (frm.doc.status === 'Accepted') {
				frm.add_custom_button(__('Training Completed'), function () {
					frm.set_value('status', 'Training Completed');
					frm.save();
				}, __('Set Status'));
			}

			if (frm.doc.status !== 'Rejected' && frm.doc.status !== 'Selected') {
				frm.add_custom_button(__('Rejected'), function () {
					frm.set_value('status', 'Rejected');
					frm.save();
				}, __('Set Status'));
			}

			// Show "Hold" button only if status is neither "Rejected" nor "Hold"
			if (frm.doc.status !== 'Hold' && frm.doc.status !== 'Rejected' && frm.doc.status !== 'Selected') {
				frm.add_custom_button(__('Hold'), function () {
					frm.set_value('status', 'Hold');
					frm.save();
				}, __('Set Status'));
			}

			if (['Hold', 'Local Enquiry Approved'].includes(frm.doc.status)) {
				frm.add_custom_button(__('Selected'), function () {
					frm.set_value('status', 'Selected');
					frm.save();
				}, __('Set Status'));
			}

			if (frm.doc.status === 'Open') {
				frm.add_custom_button(__('Shortlisted'), function () {
					frm.set_value('status', 'Shortlisted');
					frm.save();
				}, __('Set Status'));
			}

			if (frm.doc.status == 'Interview Completed') {
				frm.add_custom_button(__('Shortlisted from Interview'), function() {
					frm.set_value('status', 'Shortlisted from Interview');
					frm.save();
				}, __('Set Status'));
			}
		}
	}
}

frappe.ui.form.on('Applicant Interview Round', {
	create_interview: function (frm, cdt, cdn) {
		let row = locals[cdt][cdn];
		// Check if the form is dirty
		if (frm.is_dirty()) {
			frappe.msgprint(__('Please save your changes before creating or viewing an interview.'));
			return;
		}
		// View created Interviews. If no Interview exists, open a new unsaved Interview form with pre-filled values
		if (row.interview_reference) {
			frappe.set_route('Form', 'Interview', row.interview_reference);
		}
		else {
			if (!["Document Uploaded", "Interview Scheduled", "Interview Ongoing"].includes(frm.doc.status)){
				frappe.msgprint(__('Please upload the required documents before creating or viewing an interview.'));
				return;
			}
			frappe.model.with_doctype('Interview', function () {
				let new_interview = frappe.model.get_new_doc('Interview');
				new_interview.job_applicant = frm.doc.name;
				new_interview.job_title = frm.doc.job_opening;
				new_interview.interview_round = row.interview_round;
				new_interview.scheduled_on = frappe.datetime.now_date();
				new_interview.status = 'Pending';
				frappe.set_route('Form', 'Interview', new_interview.name);
			});
		}
	}
});